src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs

```diff
@@ -0,0 +1,134 @@
+
+        protected override string HostCore => HttpContext.Request.Host.Value;
+
+        protected override string PeerCore
+        {
+            get
+            {
+                if (_peer == null)
+                {
+                    var connection = HttpContext.Connection;
+                    if (connection.RemoteIpAddress != null)
+                    {
+                        _peer = (connection.RemoteIpAddress.AddressFamily == AddressFamily.InterNetwork ? "ipv4:" : "ipv6:") + connection.RemoteIpAddress + ":" + connection.RemotePort;
+                    }
+                }
+
+                return _peer;
+            }
+        }
+
+        // TODO(JunTaoLuo, JamesNK): implement this
+        protected override DateTime DeadlineCore => throw new NotImplementedException();
+
+        protected override Metadata RequestHeadersCore
+        {
+            get
+            {
+                if (_requestHeaders == null)
+                {
+                    _requestHeaders = new Metadata();
+
+                    foreach (var header in HttpContext.Request.Headers)
+                    {
+                        if (!header.Key.StartsWith(':'))
+                        {
+                            _requestHeaders.Add(header.Key, header.Value);
+                        }
+                    }
+                }
+
+                return _requestHeaders;
+            }
+        }
+
+        // TODO(JunTaoLuo, JamesNK): implement this
+        protected override CancellationToken CancellationTokenCore => throw new NotImplementedException();
+
+        protected override Metadata ResponseTrailersCore
+        {
+            get
+            {
+                if (_responseTrailers == null)
+                {
+                    _responseTrailers = new Metadata();
+                }
+
+                return _responseTrailers;
+            }
+        }
+
+        protected override Status StatusCore { get; set; }
+
+        protected override WriteOptions WriteOptionsCore { get; set; }
+
+        // TODO(JunTaoLuo, JamesNK): implement this
+        protected override AuthContext AuthContextCore => throw new NotImplementedException();
+
+        // TODO(JunTaoLuo, JamesNK): implement this
+        protected override ContextPropagationToken CreatePropagationTokenCore(ContextPropagationOptions options)
+        {
+            throw new NotImplementedException();
+        }
+
+        protected override Task WriteResponseHeadersAsyncCore(Metadata responseHeaders)
+        {
+            if (responseHeaders != null)
+            {
+                foreach (var entry in responseHeaders)
+                {
+                    if (!entry.IsBinary)
+                    {
+                        // TODO(juntaoluo): what about binary headers?
+                        HttpContext.Response.Headers[entry.Key] = entry.Value;
+                    }
+                }
+            }
+
+            return HttpContext.Response.Body.FlushAsync();
```

Comment:
Is this the right way to force sending the response headers? (we need to support a scenario where server responds with headers, but response messages are sent much later).

Btw, what happens if I've already sent some payload (= a message) in the body?  In that case this method should give an error.

Also, in Grpc.Core, you can only write the response headers once. Second attempt gives an error. Does this behave the same?

Reply:
There's a few discrepancies in behaviour but this is the closest we can get with the current APIs available.

Yes, this allows the user to send the headers without sending any body since the Body may be empty. The current implementation is that any write to the body is flushed which automatically sends headers if they have not been sent yet. There's no scenario where the body can be sent without the headers. Currently flush is going to no-op if the headers has already been sent. How important is it to throw an error instead? I need to think about how this might be achieved on our side as it's not immediately obvious how this can be achieved.

Reply:
Actually I lied. there's a super easy way to do this. `HttpContext.Response.HasStarted` I'll still addresss this in a followup PR though.